File: Week8/Python Behave/Behave Demo/features/steps/login_steps.py
from behave import given, when, then
from selenium import webdriver
from selenium.webdriver.common.by import By

BASE_URL = "https://the-internet.herokuapp.com/login"


@given("the application is running")
def step_impl(context):
    # Initialize the driver if not already done
    if not hasattr(context, 'driver'):
        options = webdriver.ChromeOptions()
        options.add_argument("--maximize-window")
        context.driver = webdriver.Chrome(options=options)
        context.driver.implicitly_wait(5)

@given("the test database is already seeded with users")
def step_impl(context):
    # For a UI demo, we assume the environment is ready
    pass

@given("the user is on the login page")
def step_impl(context):
    context.driver.get(BASE_URL)

@when('the user enters username "{username}"')
def step_impl(context, username):
    context.driver.find_element(By.ID, "username").send_keys(username)

@when('the user enters password "{password}"')
def step_impl(context, password):
    context.driver.find_element(By.ID, "password").send_keys(password)

@when("the user clicks the login button")
def step_impl(context):
    context.driver.find_element(By.CSS_SELECTOR, "button[type='submit']").click()

@then("the user should be redirected to the secure area")
def step_impl(context):
    assert "/secure" in context.driver.current_url

@then('the the page should display a message containing "{message}"')
def step_impl(context, message):
    flash_message = context.driver.find_element(By.ID, "flash").text
    assert message in flash_message

File: Week8/Java Cucumber/Exercises/FirstCucumberProject/src/test/java/com/revature/cucumber/stepdefinitions/LoginSteps.java
package com.revature.cucumber.stepdefinitions;

import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.And;
import io.cucumber.java.After;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import io.github.bonigarcia.wdm.WebDriverManager;

import static org.junit.jupiter.api.Assertions.*;
import java.time.Duration;

public class LoginSteps {

    private WebDriver driver;
    private static final String BASE_URL = "https://the-internet.herokuapp.com";
    private WebDriverWait wait;

    @Given("the user is on the login page")
    public void theUserIsOnTheLoginPage() {
        WebDriverManager.chromedriver().setup();
        driver = new ChromeDriver();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        driver.get(BASE_URL + "/login");
    }

    @When("the user enters username {string}")
    public void theUserEntersUsername(String username) {
        WebElement usernameField = driver.findElement(By.id("username"));
        usernameField.sendKeys(username);
    }

    @When("the user enters password {string}")
    public void theUserEntersPassword(String password) {
        WebElement passwordField = driver.findElement(By.id("password"));
        passwordField.sendKeys(password);
    }

    @When("the user clicks the login button")
    public void theUserClicksTheLoginButton() {
        WebElement loginButton = driver.findElement(By.className("radius"));
        loginButton.click();
    }

    @Then("the user should be redirected to the secure area")
    public void theUserShouldBeRedirectedToTheSecureArea() {
        wait.until(ExpectedConditions.urlContains("/secure"));
        assertTrue(driver.getCurrentUrl().contains("/secure"),
                "User was not redirected to secure area");
    }

    @Then("the user should see a success message containing {string}")
    public void theUserShouldSeeSuccessMessageContaining(String expectedMessage) {
        WebElement actualMessage = wait.until(ExpectedConditions.presenceOfElementLocated(By.id("flash-messages")));
        assertTrue(actualMessage.getText().contains(expectedMessage),
                "Actual message does not contain expected text");
    }

    @Then("the user should remain on the login page")
    public void theUserShouldRemainOnTheLoginPage() {
        assertTrue(driver.getCurrentUrl().contains("/login"),
                "User did not remain on the login page");
    }

    @Then("the user should see an error message containing {string}")
    public void theUserShouldSeeErrorMessageContaining(String expectedMessage) {
        WebElement actualMessage = wait.until(ExpectedConditions.presenceOfElementLocated(By.id("flash-messages")));
        assertTrue(actualMessage.getText().contains(expectedMessage),
                "Actual error message does not contain expected text");
    }

    @After
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
}